sdk/test/trace/batch_span_processor_test.cc

```diff
@@ -0,0 +1,290 @@
+{
+    std::shared_ptr<bool> is_shutdown(new bool(false));
+    std::shared_ptr<std::vector<std::unique_ptr<sdk::trace::SpanData>>> spans_received(
+        new std::vector<std::unique_ptr<sdk::trace::SpanData>>);
+
+    auto batch_processor = GetMockProcessor(spans_received, is_shutdown);
+    const int num_spans = 3;
+
+    auto test_spans = GetTestSpans(batch_processor, num_spans);
+
+    for(int i = 0; i < num_spans; ++i){
+        batch_processor->OnEnd(std::move(test_spans->at(i)));
+    }
+
+    batch_processor->Shutdown();
+
+    EXPECT_EQ(num_spans, spans_received->size());
+    for(int i = 0; i < num_spans; ++i)
+    {
+        EXPECT_EQ("Span " + i, spans_received->at(i)->GetName());
+    }
+   
+    EXPECT_TRUE(*is_shutdown);
+}
+
+TEST_F(BatchSpanProcessorTestPeer, TestForceFlush)
+{
+    std::shared_ptr<bool> is_shutdown(new bool(false));
+    std::shared_ptr<std::vector<std::unique_ptr<sdk::trace::SpanData>>> spans_received(
+        new std::vector<std::unique_ptr<sdk::trace::SpanData>>);
+
+    auto batch_processor = GetMockProcessor(spans_received, is_shutdown);
+    const int num_spans = 2048;
+
+    auto test_spans = GetTestSpans(batch_processor, num_spans);
+
+    for(int i = 0; i < num_spans; ++i){
+        batch_processor->OnEnd(std::move(test_spans->at(i)));
+    }
+
+    // force flush
+    batch_processor->ForceFlush(std::chrono::milliseconds(5000));
+
+    EXPECT_EQ(num_spans, spans_received->size());
+    for(int i = 0; i < num_spans; ++i)
+    {
+        EXPECT_EQ("Span " + i, spans_received->at(i)->GetName());
+    } 
+
+    // Create some more spans to make sure that the processor still works
+    auto more_test_spans = GetTestSpans(batch_processor, num_spans);
+    for(int i = 0; i < num_spans; ++i){
+        batch_processor->OnEnd(std::move(more_test_spans->at(i)));
+    }
+
+    batch_processor->ForceFlush(std::chrono::milliseconds(5000));
+
+    EXPECT_EQ(num_spans*2, spans_received->size());
+    for(int i = 0; i < num_spans; ++i)
+    {
+        EXPECT_EQ("Span " + i%num_spans, spans_received->at(i)->GetName());
+    } 
+}
+
+// TODO : Discuss and revise this test (maybe ForceFlush can return a bool)
+TEST_F(BatchSpanProcessorTestPeer, TestForceFlushTimeout)
+{
+    std::shared_ptr<bool> is_shutdown(new bool(false));
+    std::shared_ptr<std::vector<std::unique_ptr<sdk::trace::SpanData>>> spans_received(
+        new std::vector<std::unique_ptr<sdk::trace::SpanData>>);
+
+    const std::chrono::milliseconds export_delay(400);
+    const int num_spans = 3;
+
+    auto batch_processor = GetMockProcessor(spans_received, is_shutdown, export_delay);
+
+    auto test_spans = GetTestSpans(batch_processor, num_spans);
+
+    for(int i = 0; i < num_spans; ++i){
+        batch_processor->OnEnd(std::move(test_spans->at(i)));
+    }
+
+    // force flush
+    batch_processor->ForceFlush(std::chrono::milliseconds(300));
+
+    // TODO: Discuss what to expect here. I am not so sure on this.
```

Comment:
Please let me know your thoughts on this test, @IlyaKobelevskiy @reyang @pyohannes.

Reply:
Alright. Thanks!